sql-to-dbsp-compiler/lib/sqllib/src/variant.rs

```diff
@@ -387,18 +551,230 @@ where
     }
 }
 
-#[cfg(test)]
-mod test {
-    use super::Variant;
-    use dbsp::RootCircuit;
+//////////////////// Reverse conversions Variant -> T
 
-    #[test]
-    fn circuit_accepts_variant() {
-        let (_circuit, _input_handle) = RootCircuit::build(move |circuit| {
-            let (_stream, input_handle) = circuit.add_input_zset::<Variant>();
-            Ok(input_handle)
-        })
-        .unwrap();
+#[derive(Error, Debug)]
+pub enum MyError {
+    #[error("{0}")]
+    CustomError(String),
+}
+
+impl MyError {
+    pub fn from_string(message: String) -> Box<Self> {
+        Box::new(MyError::CustomError(message))
+    }
+
+    pub fn from_strng(message: &str) -> Box<Self> {
+        Box::new(MyError::CustomError(message.to_string()))
+    }
+}
+
+macro_rules! into {
+    ($variant:ident, $type:ty) => {
+        impl TryFrom<Variant> for $type {
+            type Error = Box<dyn Error>;
+
+            fn try_from(value: Variant) -> Result<Self, Self::Error> {
+                match value {
+                    Variant::$variant(x) => Ok(x),
+                    _ => Err(MyError::from_strng(concat!(
+                        "Not an ",
+                        stringify!($variant)
+                    ))),
+                }
+            }
+        }
+
+        impl TryFrom<Variant> for Option<$type> {
+            type Error = Box<dyn Error>;
+
+            fn try_from(value: Variant) -> Result<Self, Self::Error> {
+                match value {
+                    Variant::SqlNull => Ok(None),
+                    Variant::VariantNull => Ok(None),
+                    _ => match <$type>::try_from(value) {
+                        Ok(result) => Ok(Some(result)),
+                        Err(e) => Err(e),
+                    },
+                }
+            }
+        }
+    };
+}
+
+into!(Boolean, bool);
+into!(String, String);
+into!(Date, Date);
+into!(Time, Time);
+into!(Timestamp, Timestamp);
+into!(ShortInterval, ShortInterval);
+into!(LongInterval, LongInterval);
+into!(Geometry, GeoPoint);
+
+macro_rules! into_numeric {
+    ($type:ty, $type_name: ident) => {
+        impl TryFrom<Variant> for $type {
+            type Error = Box<dyn Error>;
+
+            ::paste::paste! {
+                fn try_from(value: Variant) -> Result<Self, Self::Error> {
+                    match value {
+                        Variant::TinyInt(x) => Ok([< cast_to_ $type_name _i8>](x)),
```

Comment:
many of these can panic, can't they? I think we should use non-panicking versions here